string_manipulation.rb

```diff
@@ -1,32 +1,215 @@
+  end
+
+  (word_begin...words_len).each do |idx|
+    char = my_words[idx]
+
+    # check for whitespace or end of string
+    if char == " " || idx == words_len - 1
+      # this is the end of the word; store idx - 1 or idx if end of string
+        word_end = idx == words_len - 1 ? idx : idx - 1
+
+        # reverse the found word in place
+        while word_begin < word_end
+          temp = my_words[word_begin]
+          my_words[word_begin] = my_words[word_end]
+          my_words[word_end] = temp
+
+          word_begin += 1
+          word_end -= 1
+        end
+
+        # set word_begin to be the current idx + 1
+        word_begin = idx + 1
+    end
+  end
+
+  return my_words
 end
 
 # A method to reverse the words in a sentence, in place.
 def reverse_sentence(my_sentence)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  full_reverse = string_reverse(my_sentence)
+  reversed_sentence = reverse_words(full_reverse)
+
+  return reversed_sentence
 end
 
 # A method to check if the input string is a palindrome.
 # Return true if the string is a palindrome. Return false otherwise.
 def palindrome_check(my_phrase)
-  puts "NOT IMPLEMENTED"
+  # doesn't work with mixed upper/lowercase
+  # puts "NOT IMPLEMENTED"
+  phrase_len = my_phrase.length
+
+  if phrase_len <= 1
+    return true # assumes empty string counts as a palindrome
+  end
+
+  idx_start = 0
+  idx_end = my_phrase.length - 1
+
+  while idx_start < idx_end
+    # check for whitespace
+    idx_start += 1 if my_phrase[idx_start] == " "
+    idx_end -= 1 if my_phrase[idx_end] == " "
+
+    if my_phrase[idx_start] != my_phrase[idx_end]
+      return false
+    end
+
+    idx_start += 1
+    idx_end -= 1
+  end
+
   return true
 end
 
 # A method that updates the string by replacing consecutive repeating characters
 # with a number representing the frequncy. The replacement is done only if the
 # string length will get reduced by the process.
 def encode_repeating(my_string)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  # iterate through string; if current char == prev char, add to count
+  # else check count; if <= 2 (len not reduced), no changes; if > 2 (reduced)
+  # change string from (current idx - count) to current idx to current char + count
+  count = 1
+  idx = 0
+
+  while idx < my_string.length
+    # assuming going out of range throws an error
+    # check for repetition
+    if idx < my_string.length - 1 && my_string[idx] == my_string[idx + 1]
+      count += 1
+      idx += 1
+
+    elsif count > 2
+      # keep first repeated char, mutate second into count, and delete all
+      # repeated chars after that
+      first_idx = idx - count + 1
+      my_string[first_idx + 1] = count.to_s
+
+      (count - 2).times do
+        my_string.slice!(first_idx + 2)
+      end
+
+      idx = first_idx + 2
+      count = 1
+
+    else
+      idx += 1
+      count = 1
+    end
+  end
 end
 
+# def encode_repeating(my_string) # not using slice!

```

Comment:
The time complexity of this would be O(n + k^2) where k is number of characters in the sequence. e.g. for "xxxyttttgeee"  n will be 12 and k will be 5. Since slice! returns a string - the portion that's deleted but you are doing it one character at a time and hence that would be O(1) space complexity. Since only one memory storage is getting created temporarily and then discarded.

Reply:
Nice job implementing this method!